1. Using adjusted closing prices from Yahoo, calculate daily returns for CVX, COP, WMT, and TGT.  Drop NaNs so that the returns of all stocks span the same time periods.  Compute the correlation matrix.  Of the six possible pairs of stocks, which two have the highest correlations?

In [1]:
import numpy as np
import yfinance as yf
tickers = ["CVX", "COP", "WMT", "TGT"]
prices = yf.download(tickers, start="1970-01-01")["Adj Close"]
rets = prices.pct_change().dropna()
rets.corr()

[*********************100%***********************]  4 of 4 completed


COP       CVX       TGT       WMT
COP  1.000000  0.684637  0.229699  0.191767
CVX  0.684637  1.000000  0.269959  0.242137
TGT  0.229699  0.269959  1.000000  0.490002
WMT  0.191767  0.242137  0.490002  1.000000

The pair with the highest correlation is CVX and COP.  The correlation is 68.5%.

2.  Compute the annualized standard deviations of all four stocks.

In [2]:
np.sqrt(252)*rets.std()

COP    0.330272
CVX    0.269839
TGT    0.322204
WMT    0.269357
dtype: float64

3.  Compute the standard deviation of a portfolio that is 50% in CVX and 50% in COP.  Compute the standard deviation of a portfolio that is 50% in WMT and 50% in COP.  Explain why the WMT/COP portfolio has a lower risk than the CVX/COP portfolio.  Is it because WMT has less risk than CVX?

Full credit is given for using either the annualized standard deviations for this exercise or the non-annualized daily standard deviations.  The WMT/COP portfolio has lower risk because the correlation between WMT and COP is lower than the correlation between CVX and COP, so the WMT/COP combination has better diversification.

In [3]:
# CVX and COP

port_var = (
    0.5**2 * rets.CVX.var() 
    + 0.5**2 * rets.COP.var() 
    + 2*0.5*0.5*rets.CVX.std()*rets.COP.std()*rets.CVX.corr(rets.COP)
)

np.sqrt(port_var)

0.01736407171143566

In [4]:
# WMT and COP

port_var = (
    0.5**2 * rets.WMT.var() 
    + 0.5**2 * rets.COP.var() 
    + 2*0.5*0.5*rets.WMT.std()*rets.COP.std()*rets.WMT.corr(rets.COP)
)

np.sqrt(port_var)

0.014630108208680421

## Extra

Notice that we get the same answers by computing the daily portfolio returns (1/2 ret1 + 1/2 ret2) and then computing the standard deviation of the daily portfolio returns.

In [6]:
(0.5*rets.CVX + 0.5*rets.COP).std()

0.017364071711435655

In [7]:
(0.5*rets.WMT + 0.5*rets.COP).std()

0.014630108208680421